sklearn/discriminant_analysis.py

```diff
@@ -485,9 +485,10 @@ def fit(self, X, y):
             raise ValueError("unknown solver {} (valid solvers are 'svd', "
                              "'lsqr', and 'eigen').".format(self.solver))
         if self.classes_.size == 2:  # treat binary case as a special case
-            self.coef_ = np.array(self.coef_[1, :] - self.coef_[0, :], ndmin=2)
+            my_type = np.float32 if (X.dtype in [np.float32, np.int32]) else np.float64
```

Comment:
you should already have called check_X_y and/or as_float_array above. Then you can just use X.dtype to specify the dtype of arrays you allocate

Reply:
Indeed, and the check_X_y of line 430 probably needs to be changed to "dtype=[np.float32, np.float64]"